File: package/main/src/Validate/isDouble.ts
/**
 * Determines if the value is a decimal number
 * @param {unknown} x - Value to check
 * @param {boolean} [loose=true] - Whether to include string representations of decimal numbers
 * @returns boolean - True if the value is a decimal number, false otherwise
 * @example isDouble(0.1); // true
 * isDouble("0.1"); // true
 * isDouble("0.1", false); // false
 */

const isDouble = <T extends boolean = true>(
  x: unknown,
  loose: T = true as T,
): x is T extends true ? number | string : number => {
  if (loose) {
    return (
      // biome-ignore lint/suspicious/noGlobalIsFinite: ignore
      isFinite(x as number) &&
      !Number.isNaN(x) &&
      Number.isFinite(Number(x)) &&
      !Number.isInteger(Number(x))
    );
  }
  return Number.isFinite(x) && !Number.isInteger(x);
};

export { isDouble };

File: package/umt_python/src/math/is_double.py
import math


def is_double(x: object, loose: bool = True) -> bool:
    """
    Determines if the value is a decimal number.

    Args:
        x: Value to check.
        loose: Whether to include string representations of decimal numbers.

    Returns:
        True if the value is a decimal number, False otherwise.

    Example:
        >>> is_double(0.1)
        True
        >>> is_double("0.1")
        True
        >>> is_double("0.1", False)
        False
    """
    if loose:
        try:
            num = float(x)  # type: ignore[arg-type]
            return math.isfinite(num) and not float(num).is_integer()
        except (ValueError, TypeError):
            return False
    return isinstance(x, float) and math.isfinite(x) and not x.is_integer()
